politeiawww/events.go

```diff
@@ -0,0 +1,404 @@
+				b.RUnlock()
+				log.Errorf("cannot fetch author for proposal: %v", err)
+				continue
+			}
+			b.RUnlock()
+
+			err = b.emailUsersForProposalVoteStarted(&proposal, author,
+				pvs.AdminUser)
+			if err != nil {
+				log.Errorf("email all admins for new submitted proposal %v: %v",
+					token, err)
+			}
+		}
+	}()
+	b.eventManager.register(EventTypeProposalVoteStarted, ch)
+}
+
+func (b *backend) setupProposalVoteStartedLogging() {
+	ch := make(chan interface{})
+	go func() {
+		for {
+			data := <-ch
+			pvs, ok := data.(EventDataProposalVoteStarted)
+			if !ok {
+				log.Errorf("invalid event data")
+				continue
+			}
+
+			// Log the action in the admin log.
+			err := b.logAdminProposalAction(pvs.AdminUser,
+				pvs.StartVote.Vote.Token, "start vote", "")
+			if err != nil {
+				log.Errorf("could not log action to file: %v", err)
+			}
+		}
+	}()
+	b.eventManager.register(EventTypeProposalVoteStarted, ch)
+}
+
+func (b *backend) setupProposalVoteAuthorizedEmailNotification() {
+	if b.cfg.SMTP == nil {
+		return
+	}
+
+	ch := make(chan interface{})
+	go func() {
+		for {
+			data := <-ch
+			pvs, ok := data.(EventDataProposalVoteAuthorized)
+			if !ok {
+				log.Errorf("invalid event data")
+				continue
+			}
+
+			token := pvs.AuthorizeVote.Token
+			p, err := b.getInventoryRecord(token)
+			if err != nil {
+				log.Errorf("proposal not found: %v", err)
+				continue
+			}
+			proposal := b._convertPropFromInventoryRecord(p)
+
+			err = b.emailAdminsForProposalVoteAuthorized(&proposal, pvs.User)
+			if err != nil {
+				log.Errorf("email all admins for new submitted proposal %v: %v",
+					token, err)
+			}
+		}
+	}()
+	b.eventManager.register(EventTypeProposalVoteAuthorized, ch)
+}
+
+func (b *backend) setupUserManageLogging() {
+	ch := make(chan interface{})
+	go func() {
+		for {
+			data := <-ch
+			ue, ok := data.(EventDataUserManage)
+			if !ok {
+				log.Errorf("invalid event data")
+				continue
+			}
+
+			// Log the action in the admin log.
+			err := b.logAdminUserAction(ue.AdminUser, ue.User,
+				ue.ManageUser.Action, ue.ManageUser.Reason)
+			if err != nil {
+				log.Errorf("could not log action to file: %v", err)
+			}
+		}
+	}()
+	b.eventManager.register(EventTypeUserManage, ch)
+}
+
+// register adds a listener channel for the given event type.
+//
+// This function must be called WITH the mutex held.
+func (e *EventManager) register(eventType EventT, listener chan interface{}) {
+	if e.Listeners == nil {
+		e.Listeners = make(map[EventT][]chan interface{})
+	}
+
+	if _, ok := e.Listeners[eventType]; ok {
+		e.Listeners[eventType] = append(e.Listeners[eventType], listener)
+	} else {
+		e.Listeners[eventType] = []chan interface{}{listener}
+	}
+}
+
+// unregister removes the given listener channel for the given event type.
+//
+// This function must be called WITH the mutex held.
+func (e *EventManager) unregister(eventType EventT, listener chan interface{}) {
+	if _, ok := e.Listeners[eventType]; ok {
```

Comment:
Style nit, but it's almost always preferred to keep the indentation to the right to a minimum and exit early.  e.g:

```Go
listeners, ok := e.Listeners[eventType]
if !ok {
	return
}

for l := range listeners {
	if l == listener { ...
	}
}
```

It results in easier to follow code that does not have the real meat of the logic buried deeply in conditional statements.  For example, in this case, it is obvious that no listeners means we're done and there is nothing to do.  Otherwise, you have to scan all the down to the closing brace to make sure there is nothing else.  This particular function is small enough that it's not too bad, but in longer functions, it makes a big difference in readability.